pddlgym/backend/state.py

```diff
@@ -0,0 +1,274 @@
+        """
+        Updates a predicate in the state with a new value.
+
+        Args:
+            predicate (Predicate): The predicate to update.
+            value (bool): The new value of the predicate.
+        """
+        if predicate not in self.predicates:
+            raise ValueError("Predicate {} is not in the state.".format(predicate))
+        self.predicates[predicate] = value
+    
+    def update_special_effect(self, special_effect, obj, args):
+        """
+        Updates a special effect in the state.
+        If the special effect is not in the state, it is added.
+        The special effect is then updated as an active update.
+
+        Args:
+            special_effect (SpecialEffect): The special effect to update.
+            obj (Object): The object to update the special effect for.
+            args (Dictionary[Object, Object]): The arguments for the special
+                effect.
+        """
+        new_special_effect = special_effect.copy(obj, args)
+        if new_special_effect not in self.special_effects:
+            self.special_effects.append(new_special_effect)
+        current = self.special_effects[self.special_effects.index(new_special_effect)]
+        current.update(self, active=True)
+        if current.completed:
+                self.special_effects.remove(current)
+
+    def check_goal(self):
+        """
+        Checks if the state satisfies the goal.
+
+        Returns:
+            bool: True if the state satisfies the goal, False otherwise.
+        """
+        satisfied = False
+        for goal_set in self.goal:
+            for goal in goal_set:
+                if not self.check_predicate(goal):
+                    break
+            else:
+                satisfied = True
+                break
+        return satisfied
+        
+    
+    def get_valid_actions(self):
+        """
+        Gets all valid actions for the state.
+
+        Returns:
+            valid_actions (Dictionary[Action, Dictionary[Object, Object]]): A
+                dictionary of valid actions for the state. The keys are the
+                actions, and the values are the arguments for the actions.
+        """
+        valid_actions = {}
+
+        for action, args in self.actions.items():
+            valid_actions[action] = []
+            for arg in args:
+                if action.check_if_valid(self, arg):
+                    valid_actions[action].append(arg)
+
+        return valid_actions
+
+    def step(self, action, args):
+        """
+        Steps the state forward by applying the effects of the action.
+
+        Args:
+            action (Action): The action to apply the effects of.
+            args (Dictionary[str, Object]): The arguments for the action.
+        
+        Returns:
+            new_state (State): The successor state.
+            done (bool): True if the goal is reached, False otherwise.
+        """
+        for special_effect in self.special_effects:
```

Comment:
If you move this special effect code after taking an action, it takes care of the issue where cooking takes 1 more timestep than in the configuration. Cutting seems to work as originally expected as well. Test for yourself